pkg/devfile/adapters/kubernetes/component/status.go

```diff
@@ -0,0 +1,380 @@
+package component
+
+import (
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+
+	"github.com/openshift/odo/pkg/devfile/adapters/common"
+	"github.com/openshift/odo/pkg/exec"
+	"github.com/openshift/odo/pkg/machineoutput"
+)
+
+// KubernetesDeploymentStatus is a simplified representation of the component's cluster resources
+type KubernetesDeploymentStatus struct {
+	DeploymentUID types.UID
+	ReplicaSetUID types.UID
+	Pods          []*corev1.Pod
+}
+
+// KubernetesPodStatus is a representation of corev1.Pod, but only containing the fields we are interested in (for later marshalling to JSON)
+type KubernetesPodStatus struct {
+	Name           string
+	UID            string
+	Phase          string
+	Labels         map[string]string
+	StartTime      *time.Time
+	Containers     []corev1.ContainerStatus
+	InitContainers []corev1.ContainerStatus
+}
+
+// Find the pod for the component and convert to KubernetesDeploymentStatus
+func (a Adapter) getDeploymentStatus() (*KubernetesDeploymentStatus, error) {
+
+	// 1) Retrieve the deployment
+	deployment, err := a.Client.KubeClient.AppsV1().Deployments(a.Client.Namespace).Get(a.ComponentName, metav1.GetOptions{})
```

Comment:
`getDeploymentStatus` should, at least as far as my understanding of k8s is consistent with how k8s actually works, be more resilient to failure than the use of those methods. My hope (and expectation) forthis logic is that it would be more likely to find the most up-to-date pod in various states of transition (multiple replica sets, multiple pods in various states, etc).